### PR TITLE
Update responses section with controllers metadata information

### PR DESCRIPTION
**Pull request by mikekistler**
Update responses section to describe how to add OpenAPI response metadata for controller-based apps.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/fundamentals/openapi/aspnetcore-openapi.md](https://github.com/dotnet/AspNetCore.Docs/blob/426f73e446c0ecf7ea0717f332c6b916a6626d4e/aspnetcore/fundamentals/openapi/aspnetcore-openapi.md) | [aspnetcore/fundamentals/openapi/aspnetcore-openapi](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/openapi/aspnetcore-openapi?branch=pr-en-us-33739) |

### CODE REVIEW

**Comment by mikekistler** [comment]
@Rick-Anderson @tdykstra @captainsafia I think this is ready for review / feedback.

**Review by Rick-Anderson — APPROVED** [review]
Check my changes, I might have changed the meaning in some.

I'll do a PR after this merges to wordsmith from `##### [Controllers](#tab/controllers)` on.

File: aspnetcore/fundamentals/openapi/aspnetcore-openapi.md [hunk 1]
```diff
@@ -279,21 +279,43 @@ When no explicit annotation is provided, the framework attempts to determine the
...
+* the content-type for the specified or inferred response body is "application/json".
+
+Similar to controller-based apps, there is no build-time validation of the OpenAPI metadata specified with the <xref:Microsoft.AspNetCore.Http.OpenApiRouteHandlerBuilderExtensions.Produces%2A> extension method or <xref:Microsoft.AspNetCore.Mvc.ProducesResponseTypeAttribute> attribute. For example, there is no build error issued if an endpoint returns a different status code than specified by a <xref:Microsoft.AspNetCore.Mvc.ProducesResponseTypeAttribute> attribute, or when the endpoint method returns an object of a different type than specified in the <xref:Microsoft.AspNetCore.Http.OpenApiRouteHandlerBuilderExtensions.Produces%2A> extension method.
+
+The following example illustrates using the <xref:Microsoft.AspNetCore.Http.OpenApiRouteHandlerBuilderExtensions.Produces%2A> extension method to specify the response type for an endpoint, with the default status code of 200 and default content type of `application/json`.
```
[the remark this answers is not in the document]
Comment (mikekistler) [reply aspnetcore/fundamentals/openapi/aspnetcore-openapi.md:304]: The last line has "extension method" repeated -- I'll fix that after committing all the suggestions.

File: aspnetcore/fundamentals/openapi/aspnetcore-openapi.md [hunk 2]
```diff
@@ -279,21 +279,43 @@ When no explicit annotation is provided, the framework attempts to determine the
...
 
 ```csharp
 app.MapGet("/todos", async (TodoDb db) => await db.Todos.ToListAsync())
   .Produces<IList<Todo>>();
 ```
 
+You can also use the <xref:Microsoft.AspNetCore.Mvc.ProducesResponseTypeAttribute> to add response metadata to an endpoint. Note that the attribute is applied to the route handler method, not the method invocation to create the route, as shown in the following example:
```
[the remark this answers is not in the document]
Comment (mikekistler) [reply aspnetcore/fundamentals/openapi/aspnetcore-openapi.md:311]: Also a little typo here I will fix later.

File: aspnetcore/fundamentals/openapi/aspnetcore-openapi.md [hunk 3]
```diff
@@ -326,6 +375,41 @@ If an endpoint can return different response types in different scenarios, you c
...
+
+```csharp
+[HttpPut("/todos/{id}")]
+[ProducesResponseType<Todo>(StatusCodes.Status200OK, "application/json")]
+[ProducesResponseType<Todo>(StatusCodes.Status201Created, "application/json")]
+[ProducesResponseType<ProblemDetails>(StatusCodes.Status400BadRequest, "application/problem+json")]
+public async Task<ActionResult<Todo>> CreateOrReplaceTodo(string id, Todo todo)
+```
+
+This example also illustrates how to define multiple response types for an action method, including the content type of the response body.
```
Comment (tdykstra) [inline aspnetcore/fundamentals/openapi/aspnetcore-openapi.md:410]:
```suggestion
This example also illustrates how to define multiple response types for an action method, including the content type of the response body.

```

**Review by tdykstra — APPROVED** [review]
Looks good; just a few minor issues.

File: aspnetcore/fundamentals/openapi/aspnetcore-openapi.md [hunk 4]
```diff
@@ -326,6 +375,42 @@ If an endpoint can return different response types in different scenarios, you c
...
+
+When not specified by an attribute:
+* the status code for the response defaults to 200,
+* the schema for the response body of 2xx responses may be inferred from the return type of the action method, e.g. from `T` in <xref:Microsoft.AspNetCore.Mvc.ActionResult%601>, but otherwise is considered to be not specified,
+* the schema for the response body of 4xx responses is inferred to be a problem details object,
+* the schema for the response body of 3xx and 5xx responses is considered to be not specified,
+* the content-type for the response body can be inferred from the return type of the action method and the set of output formatters.
+
+Note that there is no build-time validation of the OpenAPI metadata for responses in controller-based apps. For example, there is no build error issued if an action method returns a different status code than one specified by a <xref:Microsoft.AspNetCore.Mvc.ProducesResponseTypeAttribute> attribute, and may return an object of a different type than specified in the <xref:Microsoft.AspNetCore.Mvc.ActionResult%601> return type of the action method.
```
Comment (mikekistler) [inline aspnetcore/fundamentals/openapi/aspnetcore-openapi.md:398]: @captainsafia Please give this paragraph extra scrutiny.

File: aspnetcore/fundamentals/openapi/aspnetcore-openapi.md [hunk 5]
```diff
@@ -279,21 +279,46 @@ When no explicit annotation is provided, the framework attempts to determine the
...
+
+When not specified by one of the preceding strategies, the:
+* Status code for the response defaults to 200.
+* Schema for the response body can be inferred from the implicit or explicit return type of the endpoint method, for example, from `T` in <xref:System.Threading.Tasks.Task%601>; otherwise, it's considered to be unspecified.
+* Content-type for the specified or inferred response body is "application/json".
 
-OpenAPI supports providing a description of the responses returned from an API. Minimal APIs support three strategies for setting the response type of an endpoint:
+Similar to controller-based apps, there is no build-time validation of the OpenAPI metadata specified with the <xref:Microsoft.AspNetCore.Http.OpenApiRouteHandlerBuilderExtensions.Produces%2A> extension method or the <xref:Microsoft.AspNetCore.Mvc.ProducesResponseTypeAttribute> attribute. For example, there is no build error issued:
```
Comment (mikekistler) [inline aspnetcore/fundamentals/openapi/aspnetcore-openapi.md:302]: @captainsafia Another part to give extra scrutiny.

File: aspnetcore/fundamentals/openapi/aspnetcore-openapi.md [hunk 6]
```diff
@@ -279,21 +279,46 @@ When no explicit annotation is provided, the framework attempts to determine the
...
 #### Describe response types
 
-<!-- TODO: Restructure this section to cover both controller-based and minimal API apps -->
+OpenAPI supports providing a description of the responses returned from an API. ASP.NET Core provides several strategies for setting the response metadata of an endpoint. Response metadata that can be set includes the status code, the type of the response body, and content type(s) of a response. Responses in OpenAPI may have additional metadata, such as description, headers, links, and examples. This additional metadata can be set with a [document transformer](#use-document-transformers) or [operation transformer](#use-operation-transformers).
+
+The specific mechanisms for setting response metadata depend on the type of app being developed.
+
+##### [Minimal APIs](#tab/minimal-apis)
+
+In Minimal API apps, ASP.NET Core can extract the response metadata from extension methods on the endpoint, attributes on the route handler, and the return type of the route handler.
```
Comment (captainsafia) [inline aspnetcore/fundamentals/openapi/aspnetcore-openapi.md:288]:
```suggestion
In Minimal API apps, ASP.NET Core can extract the response metadata added by extension methods on the endpoint, attributes on the route handler, and the return type of the route handler.
```

Something to make it more clear that these strategies add the metadata to the endpoint and are not the source of metadata themselves.

File: aspnetcore/fundamentals/openapi/aspnetcore-openapi.md (hunk 5 above)
Comment (captainsafia) [reply aspnetcore/fundamentals/openapi/aspnetcore-openapi.md:302]: "build-time validation" feels like an unfamiliar turn of phrase for me, but I'm having trouble coming up with better alternatives. Perhaps this is fine given that you provide examples below.

**Pull request merged**